### ISSUE TITLE
TEDIT lets you shift-select from arbitrary parts of the expanded menu into expanded menu fields

### ISSUE THREAD

**Issue by nbriggs** [Interlisp/medley#1297]
If you open an expanded menu (say, Page Layout) then put the cursor in a fillable field you can then shift-select a line (from the left margin area) into the fillable field, and the "buttons" in the line appear to still be active - not clear whether they *really* work, but they are selectable.

**Comment by rmkaplan** [comment]
Is this in the rmk79 (new Tedit) branch, or the release (old) Tedit?

**Comment by nbriggs** [comment]
New TEdit.

**Comment by rmkaplan** [comment]
Well, same behavior in the old release version.

The lines in the menu are ordinary Tedit lines with respect to copy-select.  You can copy-select them into an exec window, you can even copy-select individual words or characters out of those lines (including the field characters).

And the open fields can also be targets for copy-select from somewhere else, e.g. if that’s where you want to get a number to fill in as the line-leading.

What’s special about the menus is that every thing except the open fields is protected—you can’t edit them or change the keywords that surround the fields.

Copy-selecting from the menus is strange, but could have its uses.

What does seem to be a non-feature/bug is the fact that (in both old and new) there is no indication to the user when it tries to apply something that doesn’t make sense, for example having the line-leading be a piece of text, not a number.  Presumably (I’ll check this in the new) it knows that it can’t install a word as a number parameter, but it should at least put something up in the promptwindow when it recognizes that the user has made that error.

**Comment by nbriggs** [comment]
Yeah, it's difficult to get rational behavior without losing some of the useful things.  It certainly makes sense to copy-select from a menu into something else, and to copy-select something into a fillable field IF it is appropriate for the field.
If you get something in the field that doesn't make sense for the field it can be difficult to get rid of it -- here I have a piece of the menu copied into the "space between cols" fill-in.  Any attempt to clear it results in a not-a-number error:

![image](https://github.com/Interlisp/medley/assets/8138841/1250de92-ff7e-468e-8ec2-b651f68658c3)

You can't shift-select only a button (or the characters in any protected text), but you can always shift-select a whole line, and once the line is in the fill-in you can't shift-select FROM it in the field (because it's protected?).  Note that the buttons are still active in the new location.

Here's an interesting case:

![image](https://github.com/Interlisp/medley/assets/8138841/d9555e0c-12b9-4410-90f6-0307b6fe2665)

With the "Hardcopy ..." line copied into the "copies" fill-in, and the fields within the field filled-in, that Hardcopy button works (modulo there being no host "xxx" in this example), but if you hit the outer Hardcopy button, it errors because an image object isn't selectable as a string.

It's all just rather weird - but I guess the real answer is "don't _do_ that!"

**Comment by rmkaplan** [comment]
This is getting tricky.  I fixed it to at least cause it to warn and ignore when a numeric fill-in doesn’t have a number, I can upload that update.

Your new example shows that when you copy from a menu to somewhere else, you might want to take away some of its properties, like convert protected to unprotected so you can edit the words you copied over, wherever they went. And with respect to the buttons, it might make sense for those image objects to have their whencopiedfn or wheninsertedfn to say DON’T, or to change to just plain text with those looks

The trickiness is that John creates prototypes of all the menus when TEDIT is loaded, and then he does a COPYTEXTSTREAM each time a menu is brought up for a new document.  So it already depends on the ability to copy every thing with all properties and behaviors intact, at least once.

I’m not sure how to recognize this second-copy situation. Maybe it could be keyed off the difference between copying a whole stream (COPYTEXTSTREAM) and copying a selection (TEDIT.COPY and copy-selecting).  Changing the protection would probably be easy.  Changing the button behavior would be harder, although perhaps each of the objects’s  buttoneventfn could refuse to act if it is not in a text object marks as a menu.

Any other ideas?

**Comment by nbriggs** [comment]
Fixes have been merged.
